File: apps/tizen/src/api/playbackUrlBuilder.ts
/**
 * Builds direct playback URLs for Xtream Codes streams.
 *
 * Port of the iOS `PlaybackURLBuilder` (Networking/PlaybackURLBuilder.swift).
 * The iOS MockFixture/demo-URL short-circuit is intentionally not ported.
 */

/** Minimal structural playlist shape needed to build playback URLs. */
export interface PlaybackPlaylist {
  serverURL: string;
  username: string;
  password: string;
}

function cleanBaseURL(serverURL: string): string {
  let baseString = serverURL.trim();

  const lower = baseString.toLowerCase();
  if (lower.indexOf("http://") !== 0 && lower.indexOf("https://") !== 0) {
    baseString = "http://" + baseString;
  }

  // Remove trailing slash
  if (baseString.charAt(baseString.length - 1) === "/") {
    baseString = baseString.slice(0, -1);
  }

  // Remove player_api.php if user entered it
  const lowered = baseString.toLowerCase();
  if (lowered.slice(-15) === "/player_api.php") {
    baseString = baseString.slice(0, -15);
  } else if (lowered.slice(-14) === "player_api.php") {
    baseString = baseString.slice(0, -14);
  }

  return baseString.split(" ").join("");
}

function authPath(playlist: PlaybackPlaylist): string {
  const u = playlist.username.trim();
  const p = playlist.password.trim();
  return u + "/" + p;
}

export class PlaybackUrlBuilder {
  private readonly playlist: PlaybackPlaylist;

  constructor(playlist: PlaybackPlaylist) {
    this.playlist = playlist;
  }

  /** Builds URL for a live stream. Xtream API direct playback format uses server/u/p/id. */
  liveUrl(streamId: number, extension?: string | null): string {
    let urlString =
      cleanBaseURL(this.playlist.serverURL) +
      "/" +
      authPath(this.playlist) +
      "/" +
      String(streamId);
    if (extension !== undefined && extension !== null && extension !== "") {
      urlString += "." + extension;
    }
    return urlString;
  }

  /** Builds URL for a VOD (Movie). */
  movieUrl(streamId: number, containerExtension?: string | null): string {
    const ext =
      containerExtension !== undefined && containerExtension !== null
        ? containerExtension
        : "mp4";
    return (
      cleanBaseURL(this.playlist.serverURL) +
      "/movie/" +
      authPath(this.playlist) +
      "/" +
      String(streamId) +
      "." +
      ext
    );
  }

  /** Builds URL for a series episode. */
  seriesUrl(streamId: string, containerExtension?: string | null): string {
    const ext =
      containerExtension !== undefined && containerExtension !== null
        ? containerExtension
        : "mp4";
    return (
      cleanBaseURL(this.playlist.serverURL) +
      "/series/" +
      authPath(this.playlist) +
      "/" +
      streamId +
      "." +
      ext
    );
  }
}

File: apps/ios/another-iptv-player/Networking/PlaybackURLBuilder.swift
import Foundation

struct PlaybackURLBuilder {
    let playlist: Playlist
    
    private var cleanBaseURL: String {
        var baseString = playlist.serverURL.trimmingCharacters(in: .whitespacesAndNewlines)
        
        if !baseString.lowercased().hasPrefix("http://") && !baseString.lowercased().hasPrefix("https://") {
            baseString = "http://\(baseString)"
        }
        
        // Remove trailing slash
        if baseString.hasSuffix("/") {
            baseString.removeLast()
        }
        
        // Remove player_api.php if user entered it
        if baseString.lowercased().hasSuffix("/player_api.php") {
            baseString = String(baseString.dropLast(15))
        } else if baseString.lowercased().hasSuffix("player_api.php") {
            baseString = String(baseString.dropLast(14))
        }
        
        return baseString.replacingOccurrences(of: " ", with: "")
    }
    
    /// Path segmenti için izinli karakterler: `urlPathAllowed` eksi segment ayırıcı `/`.
    private static let pathSegmentAllowed: CharacterSet = {
        var set = CharacterSet.urlPathAllowed
        set.remove(charactersIn: "/")
        return set
    }()

    private var authPath: String {
        // Ham birleştirme, '#'/'?'/'/'/'%' içeren hesaplarda URL'yi sessizce bozuyordu
        // (login query-item kodlamasından geçtiği için çalışıyor, oynatma çalışmıyordu).
        let u = playlist.username.trimmingCharacters(in: .whitespacesAndNewlines)
        let p = playlist.password.trimmingCharacters(in: .whitespacesAndNewlines)
        let eu = u.addingPercentEncoding(withAllowedCharacters: Self.pathSegmentAllowed) ?? u
        let ep = p.addingPercentEncoding(withAllowedCharacters: Self.pathSegmentAllowed) ?? p
        return "\(eu)/\(ep)"
    }
    
    /// Builds URL for a live stream. Xtream API direct playback format uses server/u/p/id.
    func liveURL(streamId: Int, extension: String? = nil) -> URL? {
        var urlString = "\(cleanBaseURL)/\(authPath)/\(streamId)"
        if let ext = `extension`, !ext.isEmpty {
            urlString += ".\(ext)"
        }
        return URL(string: urlString)
    }
    
    /// Builds URL for a VOD (Movie).
    func movieURL(streamId: Int, containerExtension: String?) -> URL? {
        if let mock = MockFixture.demoPlaybackURL(playlistId: playlist.id) { return mock }
        let ext = containerExtension ?? "mp4"
        let urlString = "\(cleanBaseURL)/movie/\(authPath)/\(streamId).\(ext)"
        return URL(string: urlString)
    }

    /// Builds URL for a series episode.
    func seriesURL(streamId: String, containerExtension: String?) -> URL? {
        if let mock = MockFixture.demoPlaybackURL(playlistId: playlist.id) { return mock }
        let ext = containerExtension ?? "mp4"
        let urlString = "\(cleanBaseURL)/series/\(authPath)/\(streamId).\(ext)"
        return URL(string: urlString)
    }
}
